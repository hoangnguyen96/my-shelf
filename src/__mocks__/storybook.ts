export const mockRouter = {
  pathname: "/",
  route: "/",
  query: {},
  asPath: "/",
  basePath: "",
  push: () => Promise.resolve(true),
  replace: () => Promise.resolve(true),
  reload: () => {},
  back: () => {},
  prefetch: () => Promise.resolve(),
  beforePopState: () => {},
  isFallback: false,
  isLocaleDomain: false,
  isReady: true,
  isPreview: false,
  forward: () => {},
  events: {
    on: () => {},
    off: () => {},
    emit: () => {},
  },
};
