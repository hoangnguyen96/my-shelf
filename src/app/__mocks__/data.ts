export const DATA_BOOKS = [
  {
    title: "Don't Make Me Think",
    author: "Steve Krug",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/QbQgVtG/book1.png",
    description: "description will update later",
    status: false,
    publicationYear: 2000,
    rating: 4.5,
    createdDate: "3 Sep 2024 9:07 PM",
    edition: "first",
    id: "1",
  },
  {
    title: "Rich Dad Poor Dad",
    author: "Robert T. Kiyosaki",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/9qHTtQF/book2.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1997,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:47 PM",
    edition: "second",
    id: "2",
  },
  {
    title: "The Design of Everyday Things",
    author: "Don Norman",
    category: "TextBook",
    imageUrl: "https://i.ibb.co/LnSBcCM/book5.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1988,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:51 PM",
    edition: "fourth",
    id: "3",
  },
];

export const DATA_USER = [
  {
    username: "admin",
    email: "admin@gmail.com",
    password: "$2a$10$MwSZ9vWSX7DVJwJ4SF5Mzuj2zMFjY52XM9L11gGG0j79gybowaLfq",
    isAdmin: true,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/RHMqQGr/man-1.png",
    favorites: ["3", "2", "1", "13", "5"],
    shelfBooks: ["1", "6"],
    userId: "3733403",
    id: "1",
  },
  {
    username: "hung",
    email: "hung1@gmail.com",
    password: "$2a$10$Mt6ruwOzt3nzX1.HpSt.OOzut.teXmLE74WclmkP.Nn.gzVQrlqri",
    isAdmin: false,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/88X1WfZ/avatar-default.png",
    favorites: ["1"],
    shelfBooks: ["2", "3", "1"],
    userId: "8857134",
    id: "2",
  },
];
