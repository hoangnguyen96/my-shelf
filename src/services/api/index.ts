import { MESSAGES } from "@app/constants";

const API_END_POINT = process.env.NEXT_PUBLIC_API_URL || "";

class ApiService {
  private _apiName: string;

  constructor(api: string) {
    this._apiName = api;
  }

  private async request<T>(
    method: string,
    url: string,
    body?: any,
    options?: RequestInit
  ) {
    const response = await fetch(`${this._apiName}${url}`, {
      method: method,
      headers: { "content-type": "application/json" },
      body: body ? JSON.stringify(body) : null,
      ...options,
    });

    const data = await response.json();

    if (!response.ok) {
      return {
        data: undefined,
        error: new Error(MESSAGES.NETWORK_ERROR),
      };
    }

    return {
      data: data as T,
      error: undefined,
    };
  }

  async get<T>(url: string, options?: RequestInit) {
    try {
      return await this.request<T>("GET", url, null, options);
    } catch (error) {
      throw new Error(MESSAGES.GET_ERROR);
    }
  }

  async post<T>(url: string, body: any, options?: RequestInit) {
    try {
      return await this.request<T>("POST", url, body, options);
    } catch (error) {
      throw new Error(MESSAGES.POST_ERROR);
    }
  }

  async put<T>(url: string, body: any, options?: RequestInit) {
    try {
      return await this.request<T>("PUT", url, body, options);
    } catch (error) {
      throw new Error(MESSAGES.UPDATE_ERROR);
    }
  }

  async delete<T>(url: string, options?: RequestInit) {
    try {
      return await this.request<T>("DELETE", url, null, options);
    } catch (error) {
      throw new Error(MESSAGES.DELETE_ERROR);
    }
  }
}

export const api = new ApiService(API_END_POINT);
