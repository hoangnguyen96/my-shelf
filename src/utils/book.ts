import { DEFAULT_LIMIT, TYPE_SEARCH } from "@app/constants";
import { BookType } from "@app/interface";

export const filterBooksOnShelf = (
  allBooks: BookType[],
  shelfBooks: string[]
) => allBooks.filter((item) => shelfBooks.includes(item.id));

export const filterBooksOnShelfByParams = (
  booksOnShelf: BookType[],
  type: string,
  value: string
) =>
  booksOnShelf.filter((item) =>
    type === TYPE_SEARCH.TITLE && value
      ? item.title.toLowerCase().includes(value.toLowerCase())
      : type === TYPE_SEARCH.AUTHOR && value
        ? item.author.toLowerCase().includes(value.toLowerCase())
        : item
  );

export const filterBooksFavorite = (
  allBooks: BookType[],
  favorites: string[]
) => allBooks.filter((item) => favorites.includes(item.id));

export const filterBooksFavoriteByParams = (
  booksFavorites: BookType[],
  type: string,
  value: string
) =>
  booksFavorites.filter((item) =>
    type === TYPE_SEARCH.TITLE && value
      ? item.title.toLowerCase().includes(value.toLowerCase())
      : type === TYPE_SEARCH.AUTHOR && value
        ? item.author.toLowerCase().includes(value.toLowerCase())
        : item
  );

export const dividePaginationBooks = (list: BookType[], limit?: number) => {
  return list.reduce((acc: BookType[][], _, i, self) => {
    if (!(i % (limit || DEFAULT_LIMIT))) {
      return [...acc, self.slice(i, i + (limit || DEFAULT_LIMIT))];
    }
    return acc;
  }, []);
};
