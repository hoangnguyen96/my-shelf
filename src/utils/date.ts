import { MONTHS, TIME_PERIODS } from "@app/constants";

export const formatDate = (date: Date) => {
  const hours = date.getHours();
  const minutes = date.getMinutes();
  const period = hours >= 12 ? TIME_PERIODS.PM : TIME_PERIODS.AM;
  const formattedHours = hours % 12 || 12;
  const formattedMinutes = minutes < 10 ? `0${minutes}` : minutes;

  return `${date.getDate()} ${MONTHS[date.getMonth()]} ${date.getFullYear()} ${formattedHours}:${formattedMinutes} ${period}`;
};
