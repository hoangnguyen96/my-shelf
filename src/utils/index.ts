export * from "./book";
export * from "./date";
export * from "./error";
export * from "./search";
export * from "./user";
export * from "./validators";
