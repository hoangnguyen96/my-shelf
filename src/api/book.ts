import { API_ROUTES } from "@app/constants";
import { BookType } from "@app/models";
import { HttpClient } from "@app/services";

export const getAllBook = async () => {
  try {
    const data = await HttpClient.get<BookType[]>(API_ROUTES.BOOKS);

    return data;
  } catch (error) {
    return [];
  }
};

export const getBookByParams = async (params: string) => {
  try {
    const data = await HttpClient.get<BookType[]>(
      `${API_ROUTES.BOOKS}?${params}`
    );

    return data;
  } catch (error) {
    return [];
  }
};

export const getBookById = async (id: number) => {
  try {
    const data = await HttpClient.get<BookType[]>(
      `${API_ROUTES.BOOKS}?id=${id}`
    );

    return data[0];
  } catch (error) {
    return [];
  }
};

export const addBook = async (payload: Partial<BookType>) => {
  return await HttpClient.post(API_ROUTES.BOOKS, payload);
};

export const updateBookById = async (
  id: string,
  payload: Partial<BookType>
) => {
  return await HttpClient.put(`${API_ROUTES.BOOKS}/${id}`, payload);
};

export const deleteBook = async (id: string) => {
  return await HttpClient.delete(`${API_ROUTES.BOOKS}/${id}`);
};
