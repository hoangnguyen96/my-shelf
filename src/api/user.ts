import { API_ROUTES } from "@app/constants";
import { User } from "@app/models";
import { HttpClient } from "@app/services";

export const getAllUser = async () => {
  const data = await HttpClient.get<User[]>(API_ROUTES.USER);

  return data || [];
};

export const getUserById = async (id: string) => {
  try {
    const data = await HttpClient.get<User[]>(
      `${API_ROUTES.USER}?userId=${id}`
    );

    return data[0];
  } catch (error) {
    return [];
  }
};

export const getUserByEmail = async (email: string) => {
  try {
    const user = (await HttpClient.get(
      `${API_ROUTES.USER}?email=${email}`
    )) as User[];

    return user;
  } catch (error) {
    return [];
  }
};

export const updateUserById = async (id: string, payload: Partial<User>) => {
  return await HttpClient.put(`${API_ROUTES.USER}/${id}`, payload);
};

export const addUser = async (payload: Partial<User>) => {
  return await HttpClient.post(API_ROUTES.USER, payload);
};
