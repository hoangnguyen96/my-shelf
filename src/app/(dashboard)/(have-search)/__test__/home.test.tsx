import { act, render } from "@testing-library/react";
import { useSession } from "next-auth/react";
import HomePage from "../home/page";
import { getAllBook, getUserById } from "@app/api";
import { DATA_BOOKS, DATA_USER } from "@app/app/__mocks__/data";
import * as utils from "@app/utils";

jest.mock("@app/api", () => ({
  getAllBook: jest.fn(),
  getUserById: jest.fn(),
  updateUserById: jest.fn(),
}));

jest.mock("next-auth/react", () => ({
  useSession: jest.fn(),
}));

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

jest.mock("@app/utils", () => ({
  ...jest.requireActual("@app/utils"),
  getTwelveItemData: jest.fn(),
}));

describe("HomePage", () => {
  const mockBooks = DATA_BOOKS.slice(0, 12);

  (useSession as jest.Mock).mockReturnValue({
    data: {
      user: {
        isAdmin: true,
        email: "admin@gmail.com",
        id: "3733403",
        name: "admin",
        image: "https://i.ibb.co/RHMqQGr/man-1.png",
      },
    },
    status: "authenticated",
  });

  beforeEach(() => {
    jest.clearAllMocks();
    jest.spyOn(utils, "getTwelveItemData").mockReturnValue(mockBooks);
    (getAllBook as jest.Mock).mockReturnValue({
      data: DATA_BOOKS,
    });
    (getUserById as jest.Mock).mockReturnValue({
      data: DATA_USER[0],
    });
  });

  it("Should render correctly snapshot", async () => {
    await act(async () => {
      const { container } = render(<HomePage />);
      expect(container).toMatchSnapshot();
    });
  });
});
