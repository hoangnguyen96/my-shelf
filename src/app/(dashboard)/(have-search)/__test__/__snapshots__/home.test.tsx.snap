// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Should render correctly snapshot 1`] = `<div />`;
